### PR TITLE
Restructure dependencies, add version resolution strategy.

### PR DESCRIPTION
**Pull request by artem-zinnatullin**
cc @vanniktech

### CODE REVIEW

**Comment by artem-zinnatullin** [comment]
It doesn't use [new dependency management apis](https://docs.gradle.org/4.8/userguide/dependency_locking.html) btw, so far I'm not sure that was right move for Gradle, seems to be just copy-paste from other build systems 

Extracted Gradle plugins too

File: gradle/dependency-resolution.gradle [hunk 1]
```diff
@@ -0,0 +1,25 @@
+subprojects { subProject ->
+    afterEvaluate {
```
Comment (tasomaniac) [inline gradle/dependency-resolution.gradle]: If you use `withId`, Gradle will call you when the plugin is applied. With that, you don't need `afterEvaluate`. Which means that your setup will be applied even earlier. Down-side is you need to do the following for each id.

```
project.plugins.withId('com.android.application') {
    // do the thing
}
```

Comment (artem-zinnatullin) [reply gradle/dependency-resolution.gradle]: I use similar `project.plugins.whenPluginAdded` in few scripts which triggers whenever plugin is added which is as early as possible afaik, but I'm afraid there was some lifecycle issue with this approach with dependencies

Comment (tasomaniac) [reply gradle/dependency-resolution.gradle]: I thought `resolutionStrategy` is not doing anything immediately. It is used as a strategy once dependencies are resolved. So setting the strategy as early as possible would be preferred, wouldn't it?

**Pull request merged**